### PR TITLE
Completly migrate UserPage and EditUserPage to angularjs

### PR DESCRIPTION
**Pull request by GetFire**

### CODE REVIEW

File: src/main/java/net/greatstart/controllers/UserRestController.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        return new ResponseEntity<DtoUserProfile>(HttpStatus.NOT_FOUND);
+    }
+
+    @GetMapping("/api/user/{id}")
+    public ResponseEntity<DtoUserProfile> getUserById(@PathVariable("id") long id) {
+        User user = userService.getUserById(id);
+        if (user != null) {
+            DtoUserProfile dtoUser = userMapper.fromUserToDtoProfile(user);
+            return new ResponseEntity<DtoUserProfile>(dtoUser, HttpStatus.OK);
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/UserRestController.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/net/greatstart/controllers/UserRestController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            @Valid @RequestBody DtoUserProfile dtoUser) {
+        DtoUserProfile currentDtoUser = new DtoUserProfile();
+        User currentUser = userService.getUserById(id);
+        if (currentUser != null && id == currentUser.getId()) {
+            User entityUser = userMapper.fromDtoProfileToUser(dtoUser);
+            entityUser.setPassword(currentUser.getPassword());
+            userService.updateUser(entityUser);
+            currentDtoUser = userMapper.fromUserToDtoProfile(entityUser);
+        }
+        return new ResponseEntity<DtoUserProfile>(currentDtoUser, HttpStatus.OK);
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/UserRestController.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/net/greatstart/controllers/UserRestController.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @GetMapping("/api/user/{id}")
+    public ResponseEntity<DtoUserProfile> getUserById(@PathVariable("id") long id) {
+        User user = userService.getUserById(id);
+        if (user != null) {
+            DtoUserProfile dtoUser = userMapper.fromUserToDtoProfile(user);
+            return new ResponseEntity<DtoUserProfile>(dtoUser, HttpStatus.OK);
+        }
+        return new ResponseEntity<DtoUserProfile>(HttpStatus.NOT_FOUND);
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/UserRestController.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/net/greatstart/controllers/UserRestController.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    public Principal user(Principal principal) {
+        return principal;
+    }
+
+    @GetMapping("/api/current")
+    public ResponseEntity<DtoUserProfile> getUser(Principal principal) {
+        User user = userService.getUserByEmail(principal.getName());
+        if (user != null) {
+            DtoUserProfile dtoUser = userMapper.fromUserToDtoProfile(user);
+            return new ResponseEntity<DtoUserProfile>(dtoUser, HttpStatus.OK);
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/UserRestController.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/net/greatstart/controllers/UserRestController.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        return new ResponseEntity<DtoUserProfile>(HttpStatus.NOT_FOUND);
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/controllers/UserRestController.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/net/greatstart/services/UserService.java [hunk 6]
```diff
@@ -52,6 +52,7 @@ public void deleteUser(long id) {
     }
 
     public User getUserById(long id) {
+        System.out.println("return user by id " + id);
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/services/UserService.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS106)

File: src/main/java/net/greatstart/services/UserService.java [hunk 7]
```diff
@@ -67,6 +68,7 @@ public User changeUserPassword(User user, String password) {
     }
 
     public User getUserByEmail(String email) {
+        System.out.println("find by email " + email);
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/services/UserService.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS106)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/net/greatstart/controllers/UserRestController.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+
+import javax.validation.Valid;
+import java.security.Principal;
+
+@RestController
+public class UserRestController {
```
Comment (zozich) [inline src/main/java/net/greatstart/controllers/UserRestController.java]: Word Rest is not needed in controller name. All your controller will be rest ones

File: src/main/java/net/greatstart/controllers/UserRestController.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @RequestMapping("/user")
+    public Principal user(Principal principal) {
+        return principal;
+    }
+
+    @GetMapping("/api/current")
+    public ResponseEntity<DtoUserProfile> getUser(Principal principal) {
+        User user = userService.getUserByEmail(principal.getName());
```
Comment (zozich) [inline src/main/java/net/greatstart/controllers/UserRestController.java]: All logic should be moved to a service level

File: src/main/resources/static/app/service/LoginService.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+var userService = angular.module('greatStartApp')
+    .factory('loginService', function ($rootScope, $http) {
```
Comment (zozich) [inline src/main/resources/static/app/service/LoginService.js]: Name of factories usually starts from uppercase letter - the same as for Java classes

File: src/main/resources/static/views/user/UserController.js [hunk 11]
```diff
@@ -1,18 +1,56 @@
 var UserController = angular.module('greatStartApp')
-    .controller('UserController', ['$scope', function ($scope) {
+    .controller('UserController', function ($scope, $rootScope, User, $location) {
+
         $scope.flag = false;
         $scope.myImage = '';
         $scope.myCroppedImage = '';
+        $scope.isImageChange = false;
+
+        $scope.user = angular.copy($rootScope.currentUser);
```
Comment (zozich) [inline src/main/resources/static/views/user/UserController.js]: Add toDo comment to rework this to a more direct logic. Looks like a hotfix

File: src/main/resources/static/views/user/UserController.js [hunk 12]
```diff
@@ -1,18 +1,56 @@
...
+        $scope.submit = function () {
+            if ($scope.user.id === null) {
+                console.log('Saving New User', $scope.user);
+                $scope.createUser();
+            } else {
+                if ($scope.isImageChange) {
+                    $scope.user.photo = $scope.myCroppedImage.replace(/^data:image\/[a-z]+;base64,/, "");
+                }
+                $scope.update($scope.user);
+                $location.path('/user/:id', $scope.user.id);
```
Comment (zozich) [inline src/main/resources/static/views/user/UserController.js]: You are changing location irrespective of whether you request to server was successful

File: src/test/java/net/greatstart/controllers/UserRestControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    @InjectMocks
+    private UserRestController controller;
+    private MockMvc mvc;
+    private User user;
+    private DtoUserProfile dtoUser;
+
+
+    @Before
+    public void setUp() {
+        user = new User();
```
Comment (zozich) [inline src/test/java/net/greatstart/controllers/UserRestControllerTest.java]: You don't need @Before section to do this, you can do this inline above where you declare a variable

**Review by zozich — CHANGES_REQUESTED** [review]
Please amend

File: src/main/java/net/greatstart/controllers/UserRestController.java (hunk 8 above)
Comment (GetFire) [reply src/main/java/net/greatstart/controllers/UserRestController.java]: fix

File: src/main/resources/static/app/service/LoginService.js (hunk 10 above)
Comment (GetFire) [reply src/main/resources/static/app/service/LoginService.js]: fix

File: src/main/resources/static/views/user/UserController.js (hunk 12 above)
Comment (GetFire) [reply src/main/resources/static/views/user/UserController.js]: fix

File: src/test/java/net/greatstart/controllers/UserRestControllerTest.java (hunk 13 above)
Comment (GetFire) [reply src/test/java/net/greatstart/controllers/UserRestControllerTest.java]: fix

File: src/main/java/net/greatstart/services/UserService.java [hunk 14]
```diff
@@ -16,15 +18,19 @@
...
 
     @Autowired
-    public UserService(UserDao userDao, RoleService roleService) {
+    public UserService(UserProfileMapper userMapper, UserDao userDao, RoleService roleService) {
+        this.userMapper = userMapper;
         this.userDao = userDao;
         this.roleService = roleService;
     }
 
+    //ToDo: Rewrite method for create/delete User
```
Comment (Sheptytskyid) [inline src/main/java/net/greatstart/services/UserService.java:32]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1135)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: src/main/java/net/greatstart/controllers/UserRestController.java (hunk 9 above)
Comment (GetFire) [reply src/main/java/net/greatstart/controllers/UserRestController.java]: fix

File: src/main/resources/static/views/user/UserController.js (hunk 11 above)
Comment (GetFire) [reply src/main/resources/static/views/user/UserController.js]: fix

**Review by zozich — APPROVED** [review]
OK

**Pull request merged**